File: pkg/collection/lib/src/unmodifiable_wrappers.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/**
 * Wrappers that prevent a List, Set, or Map object from being modified.
 *
 * The [Set] and [Map] wrappers allow reading from the wrapped collection,
 * but prohibit writing.
 *
 * The [List] wrapper prevents changes to the length of the wrapped list,
 * but allows changes to the contents.
 */
part of dart.pkg.collection.wrappers;

/**
 * A fixed-length list.
 *
 * A `NonGrowableListView` contains a [List] object and ensures that
 * its length does not change.
 * Methods that would change the length of the list,
 * such as [add] and [remove], throw an [UnsupportedError].
 * All other methods work directly on the underlying list.
 *
 * This class _does_ allow changes to the contents of the wrapped list.
 * You can, for example, [sort] the list.
 * Permitted operations defer to the wrapped list.
 */
class NonGrowableListView<E> extends DelegatingList<E>
                             with NonGrowableListMixin<E> {
  NonGrowableListView(List<E> listBase) : super(listBase);
}

/**
 * Mixin class that implements a throwing version of all list operations that
 * change the List's length.
 */
abstract class NonGrowableListMixin<E> implements List<E> {
  static _throw() {
    throw new UnsupportedError(
        "Cannot change the length of a fixed-length list");
  }

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void set length(int newLength) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  bool add(E value) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void addAll(Iterable<E> iterable) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void insert(int index, E element) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void insertAll(int index, Iterable<E> iterable) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  bool remove(Object value) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  E removeAt(int index) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  E removeLast() => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void removeWhere(bool test(E element)) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void retainWhere(bool test(E element)) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void removeRange(int start, int end) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void replaceRange(int start, int end, Iterable<E> iterable) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the length of the list are disallowed.
   */
  void clear() => _throw();
}

/**
 * An unmodifiable set.
 *
 * An UnmodifiableSetView contains a [Set] object and ensures
 * that it does not change.
 * Methods that would change the set,
 * such as [add] and [remove], throw an [UnsupportedError].
 * Permitted operations defer to the wrapped set.
 */
class UnmodifiableSetView<E> extends DelegatingSet<E>
                             with UnmodifiableSetMixin<E> {
  UnmodifiableSetView(Set<E> setBase) : super(setBase);
}

/**
 * Mixin class that implements a throwing version of all set operations that
 * change the Set.
 */
abstract class UnmodifiableSetMixin<E> implements Set<E> {
  _throw() {
    throw new UnsupportedError("Cannot modify an unmodifiable Set");
  }

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  bool add(E value) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void addAll(Iterable<E> elements) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  bool remove(Object value) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void removeAll(Iterable elements) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void retainAll(Iterable elements) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void removeWhere(bool test(E element)) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void retainWhere(bool test(E element)) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the set are disallowed.
   */
  void clear() => _throw();
}

/**
 * An unmodifiable map.
 *
 * An UnmodifiableMapView contains a [Map] object and ensures
 * that it does not change.
 * Methods that would change the map,
 * such as [addAll] and [remove], throw an [UnsupportedError].
 * Permitted operations defer to the wrapped map.
 */
class UnmodifiableMapView<K, V> extends DelegatingMap<K, V>
                                with UnmodifiableMapMixin<K, V> {
  UnmodifiableMapView(Map<K, V> baseMap) : super(baseMap);
}

/**
 * Mixin class that implements a throwing version of all map operations that
 * change the Map.
 */
abstract class UnmodifiableMapMixin<K, V> implements Map<K, V> {
  static _throw() {
    throw new UnsupportedError("Cannot modify an unmodifiable Map");
  }

  /**
   * Throws an [UnsupportedError];
   * operations that change the map are disallowed.
   */
  void operator []=(K key, V value) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the map are disallowed.
   */
  V putIfAbsent(K key, V ifAbsent()) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the map are disallowed.
   */
  void addAll(Map<K, V> other) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the map are disallowed.
   */
  V remove(K key) => _throw();

  /**
   * Throws an [UnsupportedError];
   * operations that change the map are disallowed.
   */
  void clear() => _throw();
}

File: pkg/dev_compiler/test/codegen/expect/collection/src/unmodifiable_wrappers.js
dart_library.library('collection/src/unmodifiable_wrappers', null, /* Imports */[
  "dart_runtime/dart",
  'dart/collection',
  'dart/core'
], /* Lazy imports */[
  'collection/wrappers'
], function(exports, dart, collection, core, wrappers) {
  'use strict';
  let dartx = dart.dartx;
  dart.export(exports, collection, ['UnmodifiableListView', 'UnmodifiableMapView'], []);
  let NonGrowableListMixin$ = dart.generic(function(E) {
    class NonGrowableListMixin extends core.Object {
      static _throw() {
        dart.throw(new core.UnsupportedError("Cannot change the length of a fixed-length list"));
      }
      set length(newLength) {
        return NonGrowableListMixin$()._throw();
      }
      add(value) {
        dart.as(value, E);
        return dart.as(NonGrowableListMixin$()._throw(), core.bool);
      }
      addAll(iterable) {
        dart.as(iterable, core.Iterable$(E));
        return NonGrowableListMixin$()._throw();
      }
      insert(index, element) {
        dart.as(element, E);
        return NonGrowableListMixin$()._throw();
      }
      insertAll(index, iterable) {
        dart.as(iterable, core.Iterable$(E));
        return NonGrowableListMixin$()._throw();
      }
      remove(value) {
        return dart.as(NonGrowableListMixin$()._throw(), core.bool);
      }
      removeAt(index) {
        return dart.as(NonGrowableListMixin$()._throw(), E);
      }
      removeLast() {
        return dart.as(NonGrowableListMixin$()._throw(), E);
      }
      removeWhere(test) {
        dart.as(test, dart.functionType(core.bool, [E]));
        return NonGrowableListMixin$()._throw();
      }
      retainWhere(test) {
        dart.as(test, dart.functionType(core.bool, [E]));
        return NonGrowableListMixin$()._throw();
      }
      removeRange(start, end) {
        return NonGrowableListMixin$()._throw();
      }
      replaceRange(start, end, iterable) {
        dart.as(iterable, core.Iterable$(E));
        return NonGrowableListMixin$()._throw();
      }
      clear() {
        return NonGrowableListMixin$()._throw();
      }
    }
    NonGrowableListMixin[dart.implements] = () => [core.List$(E)];
    dart.setSignature(NonGrowableListMixin, {
      methods: () => ({
        add: [core.bool, [E]],
        addAll: [dart.void, [core.Iterable$(E)]],
        insert: [dart.void, [core.int, E]],
        insertAll: [dart.void, [core.int, core.Iterable$(E)]],
        remove: [core.bool, [core.Object]],
        removeAt: [E, [core.int]],
        removeLast: [E, []],
        removeWhere: [dart.void, [dart.functionType(core.bool, [E])]],
        retainWhere: [dart.void, [dart.functionType(core.bool, [E])]],
        removeRange: [dart.void, [core.int, core.int]],
        replaceRange: [dart.void, [core.int, core.int, core.Iterable$(E)]],
        clear: [dart.void, []]
      }),
      statics: () => ({_throw: [dart.dynamic, []]}),
      names: ['_throw']
    });
    dart.defineExtensionMembers(NonGrowableListMixin, [
      'add',
      'addAll',
      'insert',
      'insertAll',
      'remove',
      'removeAt',
      'removeLast',
      'removeWhere',
      'retainWhere',
      'removeRange',
      'replaceRange',
      'clear',
      'length'
    ]);
    return NonGrowableListMixin;
  });
  let NonGrowableListMixin = NonGrowableListMixin$();
  let NonGrowableListView$ = dart.generic(function(E) {
    class NonGrowableListView extends dart.mixin(wrappers.DelegatingList$(E), NonGrowableListMixin$(E)) {
      NonGrowableListView(listBase) {
        super.DelegatingList(listBase);
      }
    }
    dart.setSignature(NonGrowableListView, {
      constructors: () => ({NonGrowableListView: [exports.NonGrowableListView$(E), [core.List$(E)]]})
    });
    return NonGrowableListView;
  });
  dart.defineLazyClassGeneric(exports, 'NonGrowableListView', {get: NonGrowableListView$});
  let _throw = Symbol('_throw');
  let UnmodifiableSetMixin$ = dart.generic(function(E) {
    class UnmodifiableSetMixin extends core.Object {
      [_throw]() {
        dart.throw(new core.UnsupportedError("Cannot modify an unmodifiable Set"));
      }
      add(value) {
        dart.as(value, E);
        return dart.as(this[_throw](), core.bool);
      }
      addAll(elements) {
        dart.as(elements, core.Iterable$(E));
        return this[_throw]();
      }
      remove(value) {
        return dart.as(this[_throw](), core.bool);
      }
      removeAll(elements) {
        return this[_throw]();
      }
      retainAll(elements) {
        return this[_throw]();
      }
      removeWhere(test) {
        dart.as(test, dart.functionType(core.bool, [E]));
        return this[_throw]();
      }
      retainWhere(test) {
        dart.as(test, dart.functionType(core.bool, [E]));
        return this[_throw]();
      }
      clear() {
        return this[_throw]();
      }
    }
    UnmodifiableSetMixin[dart.implements] = () => [core.Set$(E)];
    dart.setSignature(UnmodifiableSetMixin, {
      methods: () => ({
        [_throw]: [dart.dynamic, []],
        add: [core.bool, [E]],
        addAll: [dart.void, [core.Iterable$(E)]],
        remove: [core.bool, [core.Object]],
        removeAll: [dart.void, [core.Iterable]],
        retainAll: [dart.void, [core.Iterable]],
        removeWhere: [dart.void, [dart.functionType(core.bool, [E])]],
        retainWhere: [dart.void, [dart.functionType(core.bool, [E])]],
        clear: [dart.void, []]
      })
    });
    return UnmodifiableSetMixin;
  });
  let UnmodifiableSetMixin = UnmodifiableSetMixin$();
  let UnmodifiableSetView$ = dart.generic(function(E) {
    class UnmodifiableSetView extends dart.mixin(wrappers.DelegatingSet$(E), UnmodifiableSetMixin$(E)) {
      UnmodifiableSetView(setBase) {
        super.DelegatingSet(setBase);
      }
    }
    dart.setSignature(UnmodifiableSetView, {
      constructors: () => ({UnmodifiableSetView: [exports.UnmodifiableSetView$(E), [core.Set$(E)]]})
    });
    return UnmodifiableSetView;
  });
  dart.defineLazyClassGeneric(exports, 'UnmodifiableSetView', {get: UnmodifiableSetView$});
  let UnmodifiableMapMixin$ = dart.generic(function(K, V) {
    class UnmodifiableMapMixin extends core.Object {
      static _throw() {
        dart.throw(new core.UnsupportedError("Cannot modify an unmodifiable Map"));
      }
      set(key, value) {
        (() => {
          dart.as(key, K);
          dart.as(value, V);
          return UnmodifiableMapMixin$()._throw();
        })();
        return value;
      }
      putIfAbsent(key, ifAbsent) {
        dart.as(key, K);
        dart.as(ifAbsent, dart.functionType(V, []));
        return dart.as(UnmodifiableMapMixin$()._throw(), V);
      }
      addAll(other) {
        dart.as(other, core.Map$(K, V));
        return UnmodifiableMapMixin$()._throw();
      }
      remove(key) {
        return dart.as(UnmodifiableMapMixin$()._throw(), V);
      }
      clear() {
        return UnmodifiableMapMixin$()._throw();
      }
    }
    UnmodifiableMapMixin[dart.implements] = () => [core.Map$(K, V)];
    dart.setSignature(UnmodifiableMapMixin, {
      methods: () => ({
        set: [dart.void, [K, V]],
        putIfAbsent: [V, [K, dart.functionType(V, [])]],
        addAll: [dart.void, [core.Map$(K, V)]],
        remove: [V, [core.Object]],
        clear: [dart.void, []]
      }),
      statics: () => ({_throw: [dart.dynamic, []]}),
      names: ['_throw']
    });
    return UnmodifiableMapMixin;
  });
  let UnmodifiableMapMixin = UnmodifiableMapMixin$();
  // Exports:
  exports.NonGrowableListMixin$ = NonGrowableListMixin$;
  exports.NonGrowableListMixin = NonGrowableListMixin;
  exports.NonGrowableListView$ = NonGrowableListView$;
  exports.UnmodifiableSetMixin$ = UnmodifiableSetMixin$;
  exports.UnmodifiableSetMixin = UnmodifiableSetMixin;
  exports.UnmodifiableSetView$ = UnmodifiableSetView$;
  exports.UnmodifiableMapMixin$ = UnmodifiableMapMixin$;
  exports.UnmodifiableMapMixin = UnmodifiableMapMixin;
});
